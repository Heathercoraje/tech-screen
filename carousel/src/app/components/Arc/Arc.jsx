import React, { Component } from "react";
import styled from "styled-components";
import * as d3 from "d3";

export default class Arc extends Component {
  componentDidMount() {
    this.renderArc();
  }
  componentDidUpdate() {
    this.renderArc();
  }

  tau = Math.PI * 2;

  renderArc = () => {
    d3.select("svg").remove();
    const context = this.setContext();
    this.setBackground(context);
    this.setForeground(context);
  };

  setContext = () => {
    let size = this.props.size;
    return d3
      .select(this.refs.arc)
      .append("svg")
      .attr("height", size)
      .attr("width", size)
      .attr("id", "d3-arc")
      .append("g")
      .attr("transform", `translate(${size / 2}, ${size / 2})`);
  };

  arc = () => {
    let size = this.props.size;
    return d3
      .arc()
      .innerRadius(size / 2 - 10)
      .outerRadius(size / 2)
      .startAngle(0);
  };

  setBackground = context => {
    return context
      .append("path")
      .datum({ endAngle: this.tau })
      .style("fill", "transparent")
      .attr("d", this.arc());
  };

  setForeground = context => {
    const score = this.props.score;
    return context
      .append("path")
      .datum({ endAngle: this.tau * this.props.score / 700 })
      .style("fill", "powderblue")
      .attr("d", this.arc());
  };

  render() {
    return <div ref="arc" className="arc" />;
  }
}
