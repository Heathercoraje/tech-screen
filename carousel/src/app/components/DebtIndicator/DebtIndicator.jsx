import React from "react";
import styled from "styled-components";

const SmallText = styled.p`
  font-size: 1.2rem;
  color: #fff;
  margin-bottom: 5px;
  @media only screen and (min-width: 768px) {
    font-size: 1.8rem;
  }
  @media only screen and (min-width: 992px) {
    font-size: 1.8rem;
  }
`;

const BigText = styled.p`
  font-size: 4rem;
  color: lightskyblue;
  font-weight: bolder;
  margin-bottom: 5px;
  font-size: 2rem;

  @media only screen and (min-width: 330px) {
    font-size: 3rem;
  }
  @media only screen and (min-width: 992px) {
    font-size: 3.5rem;
  }
`;

const DebtIndicator = props => {
  return (
    <div
      className="container-circle"
      style={{ width: props.size, height: props.size }}
    >
      <div className="container-text">
        <SmallText>Your long term debt total</SmallText>
        <BigText> &pound; {props.debt}</BigText>
        <SmallText>Total credit limit 0</SmallText>
      </div>
    </div>
  );
};

export default DebtIndicator;
