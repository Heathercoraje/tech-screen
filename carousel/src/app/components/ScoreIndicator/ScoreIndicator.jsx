import React from "react";
import styled from "styled-components";

const SmallText = styled.p`
  font-size: 1.3rem;
  color: #fff;
  margin-bottom: 3px;
  @media only screen and (min-width: 768px) {
    font-size: 1.5rem;
  }
  @media only screen and (min-width: 992px) {
    font-size: 1.8rem;
  }
`;

const BigText = styled.p`
  font-size: 4em;
  color: lightskyblue;
  font-weight: bolder;
  margin-bottom: 3px;
  font-size: 2rem;
  @media only screen and (min-width: 300px) {
    font-size: 3rem;
  }
  @media only screen and (min-width: 992px) {
    font-size: 3.5rem;
  }
`;

const ScoreIndicator = props => {
  const size = props.size;
  return (
    <div className="container-circle" style={{ width: size, height: size }}>
      <div className="container-text">
        <SmallText>Your credit score is</SmallText>
        <BigText>{props.score}</BigText>
        <SmallText>out of 700</SmallText>
      </div>
    </div>
  );
};

export default ScoreIndicator;
