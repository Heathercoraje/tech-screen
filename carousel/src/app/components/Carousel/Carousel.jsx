import React, { Component } from "react";
import ScoreIndicator from "./../ScoreIndicator/ScoreIndicator";
import DebtIndicator from "./../DebtIndicator/DebtIndicator";
import Arc from "./../Arc/Arc";
import calculateSize from "./../../utils/calculateSize";

class Carousel extends Component {
  state = {
    index: 0,
    size: null
  };
  componentDidMount() {
    this.setState({ size: calculateSize() });
    window.setInterval(this.changeSlide, 4000);
    window.addEventListener("resize", () =>
      this.setState({
        size: calculateSize()
      })
    );
  }
  changeSlide = () => {
    let index = this.state.index === 0 ? 1 : 0;
    this.setState({ index });
  };
  renderSlides = () => {
    if (this.state.index === 0)
      return <ScoreIndicator score={this.props.score} size={this.state.size} />;
    else return <DebtIndicator debt={this.props.debt} size={this.state.size} />;
  };
  render() {
    const { score, debt } = this.props;
    if (!score) {
      return (
        <div className="loadingIndicator">
          <p>Loading . . . .</p>
        </div>
      );
    } else
      return (
        <div
          className="carousel"
          style={{
            width: this.state.size,
            height: this.state.size
          }}
        >
          <Arc size={this.state.size} score={score} />
          {this.renderSlides()}
        </div>
      );
  }
}

export default Carousel;
