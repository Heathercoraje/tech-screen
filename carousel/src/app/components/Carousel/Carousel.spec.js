import React from "react";
import { shallow } from "enzyme";

import Carousel from "./Carousel";
import Arc from "./../Arc/Arc";
import ScoreIndicator from "./../ScoreIndicator/ScoreIndicator";
import DebtIndicator from "./../DebtIndicator/DebtIndicator";
import toJson from "enzyme-to-json";

describe("Carousel renders", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Carousel />);
  });
  it("should render correctly", () => {
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
  it("without props, it should not render Arc component", () => {
    expect(wrapper.containsMatchingElement(<Arc />)).toBe(false);
  });
  it("without props, it should render `loading..` text", () => {
    expect(wrapper.contains("Loading . . . .")).toBe(true);
  });
  it("given props, it should render Arc component", () => {
    let wrapper = shallow(<Carousel score={500} />);
    expect(wrapper.containsMatchingElement(<Arc />)).toBe(true);
  });
});

describe("ScoreIndicator", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Carousel />);
  });
  it("should have index 0 as initial state", () => {
    expect(wrapper.state("index")).toBe(0);
  });
  it("given props, should render ScoreIndicator when index is 0", () => {
    let wrapper = shallow(<Carousel score={500} />);
    wrapper.setState({
      index: 0
    });
    expect(wrapper.containsMatchingElement(<ScoreIndicator />)).toBe(true);
  });
  it("given props, it should not render not DebtIndicator when index is 1", () => {
    let wrapper = shallow(<Carousel score={500} />);
    wrapper.setState({
      index: 1
    });
    expect(wrapper.containsMatchingElement(<ScoreIndicator />)).toBe(false);
  });
});

describe("DebtIndicator", () => {
  it("given props, it should render DebtIndicator when index is 1", () => {
    let wrapper = shallow(<Carousel score={500} />);
    wrapper.setState({
      index: 1
    });
    expect(wrapper.containsMatchingElement(<DebtIndicator />)).toBe(true);
  });
  it("given props, it should render not DebtIndicator when index is 0", () => {
    let wrapper = shallow(<Carousel score={500} />);
    wrapper.setState({
      index: 0
    });
    expect(wrapper.containsMatchingElement(<DebtIndicator />)).toBe(false);
  });
});
