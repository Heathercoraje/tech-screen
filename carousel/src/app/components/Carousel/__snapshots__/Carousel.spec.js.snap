// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders should render correctly 1`] = `
<div
  className="loadingIndicator"
>
  <p>
    Loading . . . .
  </p>
</div>
`;
